### DISCUSSION TITLE
Huge F_CPU dependent ADC measurement error

### DISCUSSION THREAD

**Discussion by jvasileff** [SpenceKonde/megaTinyCore#729] [Q&A] [answered]
Any idea why running a 2-series ATtiny at different clock rates would result in different values from the ADC? I'm seeing errors as large as 3% for some clock rates.

My test setup measures VCC after dividing with two 220Ω 5% resistors. I'm running at 3.3V which makes some of the higher frequencies out of spec, but the errors also happen at 8MHz.

The test sketch:

```c
void setup() {
    Serial.begin(115200);
    delay(5000);
    Serial.println("-----------------------------");
    Serial.print("F_CPU = ");
    Serial.println(F_CPU);
    Serial.println("-----------------------------");
}

void loop() {
    analogReference(INTERNAL2V048);
    // analogSampleDuration(128);
    int32_t reading = analogReadEnh(PIN_PA6, 12);
    Serial.print(reading);
    Serial.print(", ");
    
    Serial.print(reading / 4096.0 * 2.048, 3);
    Serial.println("V");

    delay(1000);
}
```

Result summary:

  | Reading (V) | Error
-- | -- | --
Multimeter | 1.624 | 0.0%
20MHz | 1.622 | -0.1%
16MHz | 1.621 | -0.2%
10MHz | 1.670 | **2.8%**
8MHz | 1.672 | **3.0%**
5MHz | 1.620 | -0.2%
4MHz | 1.648 | **1.5%**
1MHz | 1.634 | 0.6%

Raw output:

```
-----------------------------
F_CPU = 20000000
-----------------------------
3257, 1.629V
3243, 1.622V
3243, 1.622V
3243, 1.622V
3243, 1.622V

-----------------------------
F_CPU = 16000000
-----------------------------
3245, 1.623V
3242, 1.621V
3242, 1.621V
3242, 1.621V
3242, 1.621V

-----------------------------
F_CPU = 10000000
-----------------------------
3360, 1.680V
3340, 1.670V
3342, 1.671V
3340, 1.670V
3340, 1.670V

-----------------------------
F_CPU = 8000000
-----------------------------
3341, 1.671V
3344, 1.672V
3346, 1.673V
3344, 1.672V
3344, 1.672V

-----------------------------
F_CPU = 5000000
-----------------------------
3238, 1.619V
3241, 1.621V
3237, 1.619V
3241, 1.621V
3240, 1.620V

-----------------------------
F_CPU = 4000000
-----------------------------
3298, 1.649V
3296, 1.648V
3296, 1.648V
3296, 1.648V
3296, 1.648V

-----------------------------
F_CPU = 1000000
-----------------------------
3269, 1.634V
3268, 1.634V
3268, 1.634V
3268, 1.634V
3268, 1.634V
```

**Comment by jvasileff** [comment]
@SpenceKonde this may help answer part of the puzzle regarding the accuracy of the internal temperature sensor. At 5MHz & 3.3V, readings are about equal to ambient temp, but at 8MHz, they are 30F over ambient. The reading is still several degrees higher than ambient at 16MHz, but maybe that's due to the chip running hotter at higher frequencies.

**Comment by mosqu-ito** [comment]
Do you follow along with your meter to be sure the measured voltage is indeed holding solid under the different tests?

**Reply by jvasileff** [reply]
Yes - the measured voltage fluctuated much less (per the multimeter) than the values reported by the 3226s. And, the “bad” clock speeds were consistently bad throughout a number of tests as I changed things up to try to troubleshoot measurement errors on a project I’m working on.

**Comment by mosqu-ito** [comment]
Fluctuations could be very short lived.  Maybe eliminate that variable by measuring something OTHER than Vcc, say, a flashlight battery.

**Reply by jvasileff** [reply]
The 3.3V regulator I'm using as a voltage source is plenty stable for this test, and the results are very consistent across *many* samples, and as I confirmed, I kept an eye on the multimeter. I also see the same behavior measuring the output of the internal temperature sensor (ADC_TEMPERATURE), the internal VDD source (ADC_VDDDIV10), and an LMT85LP analog temperature sensor. I first noticed the problem on the LMT85LP sensor and the internal ADC_TEMPERATURE before setting up the simple test.

**Comment by jvasileff** [comment, marked as answer]
Welp... looks like the problem was bad breadboarding.

I soldered a bypass cap onto the breakout board I used for the 3226 as shown in the pic, and that cleared up most of the problem. Apparently the blue cap in the pic  - which was there before - wasn't helping at all.

![IMG_0251](https://user-images.githubusercontent.com/236050/176083384-015d986c-fbbe-427d-9dad-d46dea47cba5.jpeg)

Measurements after adding the extra 100n cap:

MHz | Reading (V)
-- | --
20 | 1.652
16 | 1.645
10 | 1.647
8 | 1.643
5 | 1.642

**Reply by mosqu-ito** [reply]
A picture is indeed worth a thousand words.  See datasheet figure 4-1.
